xarray/tests/test_dask.py

```diff
@@ -917,3 +918,176 @@ def test_dask_layers_and_dependencies():
+        return obj.chunk(new_chunks)
+
+    if isinstance(a, DataArray):
+        a = a._to_temp_dataset()
+
+    if isinstance(b, DataArray):
+        b = b._to_temp_dataset()
+
+    a = at_least_one_chunk(a.unify_chunks())
+    b = at_least_one_chunk(b.unify_chunks())
+
+    assert a.chunks == b.chunks
+
+
+def test_map_blocks_error():
+    def bad_func(darray):
+        return (darray * darray.x + 5 * darray.y)[:1, :1]
+
+    with raises_regex(ValueError, "Length of the.* has changed."):
+        xr.map_blocks(bad_func, map_da).compute()
+
+    def returns_numpy(darray):
+        return (darray * darray.x + 5 * darray.y).values
+
+    with raises_regex(ValueError, "Function must return an xarray DataArray"):
+        xr.map_blocks(returns_numpy, map_da)
+
+    with raises_regex(ValueError, "args must be"):
+        xr.map_blocks(operator.add, map_da, args=10)
+
+    with raises_regex(ValueError, "kwargs must be"):
+        xr.map_blocks(operator.add, map_da, args=[10], kwargs=[20])
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks(obj):
+    def func(obj):
+        result = obj + obj.x + 5 * obj.y
+        return result
+
+    with raise_if_dask_computes():
+        actual = xr.map_blocks(func, obj)
+    expected = func(obj).unify_chunks()
+    assert_chunks_equal(expected, actual)
+    # why is compute needed?
+    xr.testing.assert_equal(expected.compute(), actual.compute())
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks_convert_args_to_list(obj):
+    expected = obj.unify_chunks() + 10
+    with raise_if_dask_computes():
+        actual = xr.map_blocks(operator.add, obj, [10])
+    assert_chunks_equal(expected, actual)
+    # why is compute needed?
+    xr.testing.assert_equal(expected.compute(), actual.compute())
+
+
+@pytest.mark.parametrize("obj", [map_da, map_ds])
+def test_map_blocks_kwargs(obj):
+    expected = xr.full_like(obj, fill_value=np.nan)
+    with raise_if_dask_computes():
+        actual = xr.map_blocks(xr.full_like, obj, kwargs=dict(fill_value=np.nan))
+    assert_chunks_equal(expected, actual)
+    # why is compute needed?
+    xr.testing.assert_equal(expected.compute(), actual.compute())
+
+
+@pytest.mark.parametrize(
+    "func, obj",
+    [
+        [lambda x: x.to_dataset(), map_da],
+        [lambda x: x.to_array(), map_ds],
+        [lambda x: x.drop("a"), map_ds],
+        [lambda x: x.expand_dims(k=[1, 2, 3]), map_ds],
+        [lambda x: x.expand_dims(k=[1, 2, 3]), map_da],
+        [lambda x: x.isel(x=1), map_ds],
+        [lambda x: x.isel(x=1).drop("x"), map_da],
+        [lambda x: x.assign_coords(new_coord=("y", x.y * 2)), map_da],
+        [lambda x: x.astype(np.int32), map_da],
+        [lambda x: x.rename({"a": "new1", "b": "new2"}), map_ds],
+    ],
+)
+def map_blocks_transformations(func, obj, expected):
+    with raise_if_dask_computes():
+        assert_equal(xr.map_blocks(func, obj), func(obj))
```

Comment:
How is it possible that assert_equal doesn't compute anything if it is supposed to compare the contents?